### ISSUE TITLE
Floating version patterns like '1.*` are treated as ranges without a max, `1.*` is [1.*, )

### ISSUE THREAD

**Issue by anangaur** [NuGet/Home#5097]
1. Create a NETCore/NETStandard project (PackageReference).
2. Have it depend on a package (say NuGetPlayLib) that has versions 1.1.0,1.2.0,1.3.0,2.0.0
3. Now edit the csproj file to make the dependency on "1.*"
`  <ItemGroup>
     <PackageReference Include="NuGetPlayLib" Version="1.*" />
   </ItemGroup>`
4. Remove the versions 1.* from the feeds
5. Restore package dependencies for the project

**Expected:** It should throw an error as it would have if the dependency was specified as:
`  <ItemGroup>
    <PackageReference Include="NuGetPlayLib" Version="[1.0.0, 2.0.0)" />
  </ItemGroup>`
**Actual:** It resolves to version 2.0.0

**Comment by anangaur** [comment]
For reference look at the documentation:
http://blog.nuget.org/20170316/NuGet-now-fully-integrated-into-MSBuild.html

**Comment by emgarten** [comment]
Users can specify ``[1.*, 2.0.0)`` to enforce an upper bound.

**Comment by emgarten** [comment]
A floating range is treated as two ranges currently. 
1. The allowed bounds
1. A preferred version within those bounds

In the case of ``1.*`` the preferred range is ``1.MAXINT``, the nearest version to that will be used. Versions below the bounds of ``[1.0.0, )`` will be considered errors.

**Comment by anangaur** [comment]
I find this non-intuitive. The `1.*` seems like it should match to just `1.MAXINT`. Was there a requirement to match to the nearest version if the pattern match doesn't happen? I find the usecase a bit weird.

**Comment by lodejard** [comment]
The wildcard is matched to what's highest the feed first, and then acts like a normal range afterwards. So if the feed has 1.1.0,1.2.0,1.3.0,2.0.0 and you look for 1.* Then it's effectively turned to 1.3 without a maximum.

Like @emgarten mentioned then - if you have [1.*, 2.0) then it's effectively turned to [1.3, 2.0) which gives you the clamped range you were looking for.

*However* there is a problem if a wildcard automatically adds a maximum. 

Let's say you have project Q that uses library A and B, and A and B use package X. If A says X 1.4 is ideal, and B says X 2.2 is ideal, then Q gets X 2.2 which is the lowest matching package in the intersection of ranges.

But if a wildcard also implies a maximum and A says X 1.* is ideal, and B says X 2.* is ideal, then Q will get a restore error because the ranges [1.3, 2.0) and [2.0, 3.0) have no intersection.

The major version might not seem so bad, but you'd have the same problem with any non-identical wildcard... 1.5.* and 1.6.* will exclude each other, but 1.5.4 and 1.6.1 will happily select 1.6.1 for you.

**Comment by lodejard** [comment]
That said - I do think this is a case where you should get the "under-versioning" warning. Just like if I say 1.3 is ideal and restore ends up handing me 2.0 I'll get a warning, if I say 1.* is ideal and restore ends up with 2.0 I should get that same warning.

**Comment by anangaur** [comment]
@lodejard I see 2 issues with the current behavior using wild card:
1. The semantic is non intuitive. The range using braces already works right now. So if somebody wants a particular behavior like no upper bound, then they can go with no upper bound. E.g [1.0.0,] 
What will 1.* regular expression mean? Can it ever mean 2.x? 
Also since [1.0.0, 2.0.0) will never match to 2.0.0 even if there are no packages with version 1.x.x, we should have similar behavior for the *

2. If we limit 1.* to mean less than 2 equivalent to [1.0.0,2.0.0), then this also enables one to constrain the dependency range to the same major version.
E.g
1.* =  [1.0.0,2)
1.3.* = [1.3.0, 1.4.0)

**Comment by ivan-danilov** [comment]
Wow... that's completely awkward. [Docs](https://docs.microsoft.com/en-us/nuget/concepts/dependency-resolution#floating-versions) say:

> A floating or wildcard dependency version is specified with the * wildcard, as with 6.0.*. This version specification says "use the latest 6.0.x version"; 4.* means "use the latest 4.x version."

In reality, `4.*` means `[4.*, )`. That's a bummer as if package sources don't have any versions with Major=4 at all - NuGet will happily resolve `5.0.0` or whatever is available there next.

But even `[4.0.0, 4.1.0)` (assuming it resolves to the highest applicable version) does not solve the issue: `4.1.0-prerelease.label` matches that range as well. Therefore, taking the [explanations in this thread](https://github.com/NuGet/Home/issues/6434#issuecomment-358782297) into account, the correct range to specify what is expected from `4.*` by most people would be `[4.0.0, 4.1.0-0)`.

How many people would write it correctly from the first time? And what is the point in having `4.*` implemented as it is now, namely `[4.*, )`? To be honest, I struggle to find a real case when I'd like version `5.0.0` to match reference `4.*` silently.

**Comment by ivan-danilov** [comment]
In fact, even the last one is not right because of #912 as it won't consider pre-release versions if there is at least one version without pre-release label (i.e. `4.0.1-rc` won't be considered if there is `4.0.0` present in the feed). Complete bummer, seems not possible to do at all. `[4.0.0-*, 4.1.0-0)` does not work either.

**Comment by Frulfump** [comment]
Needs triage, range syntax in general needs work to align with other ecosystems like npm.

**Comment by nkolev92** [comment]
Team Triage: This is not very common (because if any version meets the requirement, it'll match without crossing the major boundary), but I think we can consider adding a warning here, or maybe reusing an existing one. 

Example: 
When 1.0.0 is specified in the project (`<PackageReference Include="package" Version="1.0.0">`) which is `>= 1.0.0` and 1.1.0 gets resolved, you get https://learn.microsoft.com/en-us/nuget/reference/errors-and-warnings/nu1603. 

We could do something similar when a floating version crosses the restriction of its version part. 
So if `1.*` is specified, but 2.0.0 is resolved, you get a warning. This would be a less aggressive way of changing the behavior. 
Then assuming minimum noise, we could even consider changing it for good.

**Comment by ivan-danilov** [comment]
@nkolev92 in our situation, we have multiple branches of our product in active support. That means 1.0.X is one family of versions and 2.0.X is another, and both are active simultaneously (we have an SLA with customers for at least 5 years of support for any major version). There are breaking changes between 1.0 and 2.0 (as implied by the major version bump), so resolving 2.0.X when 1.0.* is requested simply breaks the build. The warning won't fix it.

I understand that the situation may not be possible to solve without breaking changes, but unfortunately, the issue is not cosmetic in some cases.

**Comment by nkolev92** [comment]
@ivan-danilov 

1.0.* will only resolve 2.0.x when there are no 1.0.X matching builds on the feed. 
If you are using the same feed, then that'd suggest you have no 1.0.x builds available on your feed.

**Comment by ivan-danilov** [comment]
@nkolev92 yes, and it is a literal everyday scenario when multiple actively supported branches, which I described earlier. For example, let's imagine we have versions 1.0 and 2.0, both active (meaning that changes are made to both, whether it is fixing bugs or adding new features). There are consumers for both branches who would want to take the latest patch version of 1.0 and 2.0, respectively. Let's see what preparations for 1.1.0 would look like. It is not yet ready, so we have oncoming streams of versions with pre-release labels, e.g., for simplicity, just numbers 1.1.0-1, 1.1.0-2 etc. The consumers want to take "the latest 1.1.* available, including pre-releases". And given that 2.0.0 is already present, it seems impossible to achieve, as no _release_ versions of 1.1.0 are there yet.

**Comment by nkolev92** [comment]
> The consumers want to take "the latest 1.1.* available, including pre-releases". And given that 2.0.0 is already present, it seems impossible to achieve, as no release versions of 1.1.0 are there yet.

If you want your floating version to consider prereleases, you can specify `1.1.*-*`
This allows prerelease versions, but if a stable is available, it allows that too. 

Some examples: 

Matching prerelease gets picked.

https://nugettools.azurewebsites.net/7.3.0/find-best-version-match?versionRange=1.1.*-*&versions=1.0.0%0D%0A1.1.0%0D%0A1.1.1-preview.1%0D%0A1.1.1-rc.2%0D%0A2.0.0-rc.1%0D%0A2.0.0

Stable is preferred even 
https://nugettools.azurewebsites.net/7.3.0/find-best-version-match?versionRange=1.1.*-*&versions=1.0.0%0D%0A1.1.0%0D%0A1.1.1-preview.1%0D%0A1.1.1-rc.2%0D%0A1.1.2%0D%0A2.0.0-rc.1%0D%0A2.0.0

Finally, you could always add an upper bound for the ranges: 

https://nugettools.azurewebsites.net/7.3.0/find-best-version-match?versionRange=%5B1.1.*-*%2C+2.0.0-0%29&versions=1.0.0%0D%0A2.0.0-rc.1%0D%0A2.0.0
